include/envoy/http/filter.h

```diff
@@ -32,28 +32,42 @@ enum class FilterHeadersStatus {
   // FilterDataStatus::Continue from decodeData()/encodeData() or calling
   // continueDecoding()/continueEncoding() MUST be called if continued filter iteration is desired.
   StopIteration,
-  // Continue iteration to remaining filters, but ignore any subsequent data or trailers. This
+  // Continue headers iteration to remaining filters, but ignore any subsequent data or trailers.
+  // This
   // results in creating a header only request/response.
   // This status MUST NOT be returned by decodeHeaders() when end_stream is set to true.
   ContinueAndEndStream,
-  // Continue iteration to remaining filters, but do not end the stream.
+  // Continue headers iteration to remaining filters, but delay ending the stream. This status MUST
+  // NOT be returned when end_stream is already set to false.
   //
-  // Used when a filter wants to add a body to headers-only request/response, but this body is not
-  // readily available.
-  // This causes the headers iteration to continue, but the stream does not end.
-  // The filter is responsible to continue the stream by providing a body through
-  // injectEncodedDataToFilterChain()/injectDecodedDataToFilterChain().
+  // Used when a filter wants to add a body to a headers-only request/response, but this body is not
+  // readily available. Delaying end_stream allows the filter to add the body once it's available
+  // without stopping headers iteration.
+  //
+  // The filter is responsible to continue the stream by providing a body through calling
+  // injectDecodedDataToFilterChain()/injectEncodedDataToFilterChain(), possibly multiple times
+  // if the body needs to be divided into several chunks. The filter may need to handle
+  // watermark events when injecting a body, see:
+  // https://github.com/envoyproxy/envoy/blob/master/source/docs/flow_control.md.
+  //
+  // The last call to inject a body MUST have end_stream set to true to conclude the stream.
```

Comment:
Presumably it could also be ended with `addEncoded/DecodedTrailers`?

Reply:
Right. I will add that.

Reply:
Actually I looked at `addEncoded/DecodedTrailers` docs and code. it looks like they can only be called in `encodeData` when `end_stream` is true. So they don't apply here. Will revert the last change.

Reply:
The current API does not support adding trailers in this scenario, it would need similar `injectTrailers*` methods.

Reply:
I documented this properly.

Reply:
Thanks for looking into it!